### PR TITLE
NEW: dataclass-style settings

### PR DESCRIPTION
**Pull request by eigenein**
The package is very old and makes no use of the modern Python features. I propose that we should support dataclasses for settings – instead of plain `Settings()`. With that users would get IDEs' autocompletion features and linting checks.

This PR rewrites almost the entire implementation, yet **it keeps the backwards compatibility**, so that existing users would be able to migrate whenever they are ready.

The old tests are preserved to check the compatibility and moved to `test_legacy.py`. The new tests are the same ones but in the dataclasses style.

On top of that, the context manager is now both thread- and task-safe.

### CODE REVIEW

File: pkgsettings/__init__.py [hunk 1]
```diff
@@ -1,3 +1,3 @@
-from .pkgsettings import DuplicateConfigureWarning, PrefixedSettings, Settings, SimpleSettings
+from .pkgsettings import DuplicateConfigureWarning, PrefixedSettings, Settings
 
-__all__ = ["DuplicateConfigureWarning", "PrefixedSettings", "Settings", "SimpleSettings"]
```
Comment (eigenein) [inline pkgsettings/__init__.py:3]: This is may be considered a breaking change, however, no public API exposes parameters or return values of `SimpleSettings` directly. So, I believe it's not breaking

**Comment by woile** [comment]
I like this idea in principle, though I still haven't reviewed the code, my question is: why not recommend [pydantic BaseSettings](https://docs.pydantic.dev/usage/settings/) ? Pydantic is usually present and we could remove this packages as a dependency, meaning less maintenance burden for us.

File: pkgsettings/__init__.py (hunk 1 above)
Comment (woile) [reply pkgsettings/__init__.py:3]: AFAIK `__all__` is for doing `from X import *`, if someone is doing that in their repo... better to break it 🤣

**Review by woile — APPROVED** [review]
LGTM. Does it have suport for `.env` file (and custom files like `example.env`, `local.env`, etc)? If not, I think that should be the next step.

**Pull request closed without merging**